plugins/bundle/plugin.go

```diff
@@ -0,0 +1,282 @@
+// Copyright 2018 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+// Package bundle implements bundle downloading.
+package bundle
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"time"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/bundle"
+	"github.com/open-policy-agent/opa/plugins"
+	"github.com/open-policy-agent/opa/storage"
+	"github.com/open-policy-agent/opa/util"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// min amount of time to wait following a failure
+	minRetryDelay = time.Millisecond * 100
+)
+
+var (
+	defaultMinDelay = int64(time.Second * 60)
+	defaultMaxDelay = int64(time.Second * 120)
+)
+
+// PollingConfig represents configuration for the plugin's polling behaviour.
+type PollingConfig struct {
+	MinDelaySeconds *int64 `json:"min_delay_seconds,omitempty"` // min amount of time to wait between successful poll attempts
+	MaxDelaySeconds *int64 `json:"max_delay_seconds,omitempty"` // max amount of time to wait between poll attempts
+}
+
+// Config represents configuration the plguin.
+type Config struct {
+	Name    string        `json:"name"`
+	Service string        `json:"service"`
+	Polling PollingConfig `json:"polling"`
+}
+
+func (c *Config) validateAndInjectDefaults(services []string) error {
+
+	if c.Name == "" {
+		return fmt.Errorf("invalid bundle name %q", c.Name)
+	}
+
+	found := false
+
+	for _, svc := range services {
+		if svc == c.Service {
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		return fmt.Errorf("invalid service name %q in bundle %q", c.Service, c.Name)
+	}
+
+	if c.Polling.MinDelaySeconds == nil && c.Polling.MaxDelaySeconds == nil {
+		c.Polling.MinDelaySeconds = &defaultMinDelay
+		c.Polling.MaxDelaySeconds = &defaultMaxDelay
+	} else if c.Polling.MinDelaySeconds != nil && c.Polling.MaxDelaySeconds == nil {
+		c.Polling.MaxDelaySeconds = c.Polling.MinDelaySeconds
+	} else if c.Polling.MinDelaySeconds == nil && c.Polling.MaxDelaySeconds != nil {
+		c.Polling.MinDelaySeconds = c.Polling.MaxDelaySeconds
```

Comment:
Why are the min and max delays equal ? Shouldn't one of them be set to the default values ? If they are equal how would line 147 work ?

Reply:
Thanks. This is buggy. I'm going to amend with better testing and make the min/max behaviour clearer.

If user only sets min (or max) the other gets set to the default unless that would violate the max>=min constraint in which case min=max.

WDYT?

Reply:
Added better testing and fixed the delay calculation.

Reply:
For ex. if user sets MinDelaySeconds = 150 and does not set MaxDelaySeconds.  Given,
defaultMinDelaySeconds = int64(60) and defaultMaxDelaySeconds = int64(120).

In this case, min will be set to 120 and max will be set to 150. Is this what you mean ?

Reply:
In that case, min and max would both be set to 150.